### PR TITLE
Implement PostBeanProcessor 

### PR DESCRIPTION
**Pull request by OleksandRomaniuk**
Finally it work with email sending

### CODE REVIEW

File: src/main/kotlin/com/example/ajaxproject/pbp/NotificationAnnotationBeanPostProcessor.kt [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+@Suppress("SpreadOperator")
+class DeviceAuthorizationInvocationHandler(
+    private val bean: Any,
+    private val groupChatRoomRepository: GroupChatRoomRepository,
+    private val emailSenderService: EmailSenderService,
+    private val originalBean: KClass<*>
+) : InvocationHandler {
+    override fun invoke(proxy: Any, method: Method, args: Array<out Any>?): Any? {
```
Comment (ajax-gnatiuk-s) [inline src/main/kotlin/com/example/ajaxproject/pbp/NotificationAnnotationBeanPostProcessor.kt:60]:
```suggestion
) : InvocationHandler {

    override fun invoke(proxy: Any, method: Method, args: Array<out Any>?): Any? {
```

File: src/main/kotlin/com/example/ajaxproject/service/GroupChatServiceImpl.kt [hunk 2]
```diff
@@ -120,4 +119,12 @@ class GroupChatServiceImpl(
             throw WrongActionException("Admin cant leave a chat")
         }
     }
+    fun toResponseDto(chatMessage: GroupChatMessage): GroupChatMessageResponse {
```
Comment (ajax-gnatiuk-s) [inline src/main/kotlin/com/example/ajaxproject/service/GroupChatServiceImpl.kt:122]:
```suggestion

    fun toResponseDto(chatMessage: GroupChatMessage): GroupChatMessageResponse {
```

File: src/main/resources/application.yml [hunk 3]
```diff
@@ -7,6 +7,24 @@ spring:
...
+    port: 587
+    properties:
+      mail:
+        smtp:
+          starttls:
+            required: true
+
+
+  main:
+    allow-bean-definition-overriding: true
```
Comment (ajax-gnatiuk-s) [inline src/main/resources/application.yml:24]: why do you need this?

**Review by ajax-gnatiuk-s — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/example/ajaxproject/service/UserServiceImpl.kt [hunk 4]
```diff
@@ -5,20 +5,27 @@ import com.example.ajaxproject.exeption.NotFoundException
...
+    private val userRepository: UserRepository
 ) : UserService {
 
     override fun createUser(userDTO: UserDTO): User {
-        val user = userMapper.toEntity(userDTO)
+        val user = toEntity(userDTO)
         return userRepository.save(user)
     }
 
+    fun toEntity(userDTO: UserDTO): User {
```
[the remark this answers is not in the document]
Comment (OleksandRomaniuk) [reply src/main/kotlin/com/example/ajaxproject/service/UserServiceImpl.kt:21]: did it

File: src/main/resources/application.yml (hunk 3 above)
Comment (OleksandRomaniuk) [reply src/main/resources/application.yml:24]: Actually I need it when i wrote email service , but now its look like it work without it

File: src/main/kotlin/com/example/ajaxproject/pbp/NotificationAnnotationBeanPostProcessor.kt [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        return result
+    }
+
+    private fun createNotification(groupChatDTO: GroupChatDTO) {
+
+        val chatName = groupChatRoomRepository.findChatRoom(groupChatDTO.chatId).chatName
+
+        val userList = groupChatRoomRepository.findById(ObjectId(groupChatDTO.chatId)).get().chatMembers
+
+        logger.info("Emails sent to users: {}", userList)
```
[the remark this answers is not in the document]
Comment (OleksandRomaniuk) [reply src/main/kotlin/com/example/ajaxproject/pbp/NotificationAnnotationBeanPostProcessor.kt:83]: Make sense . Move it to new line

File: src/main/kotlin/com/example/ajaxproject/service/GroupChatServiceImpl.kt (hunk 2 above)
Comment (ajax-gnatiuk-s) [reply src/main/kotlin/com/example/ajaxproject/service/GroupChatServiceImpl.kt:123]: why this discussion was silently resolved?

You have not applied the suggested changes

**Review by ajax-gnatiuk-s — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/example/ajaxproject/service/EmailSenderServiceImpl.kt [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    }
+    private fun generateMailMessage(emailDTO: EmailDTO): MimeMessage {
+
+        val helper = MimeMessageHelper(javaMailSender.createMimeMessage())
+        helper.setFrom(emailDTO.from)
+        helper.setTo(emailDTO.to)
+        helper.setSubject(emailDTO.subject!!)
+        helper.setText(emailDTO.body, true)
+
+        return helper.mimeMessage
```
Comment (ajax-gnatiuk-s) [inline src/main/kotlin/com/example/ajaxproject/service/EmailSenderServiceImpl.kt:35]:
```suggestion
        return MimeMessageHelper(javaMailSender.createMimeMessage()).apply {
            setFrom(emailDTO.from)
            setTo(emailDTO.to)
            setSubject(emailDTO.subject!!)
            setText(emailDTO.body, true)            
        }.mimeMessage
```

File: src/main/kotlin/com/example/ajaxproject/service/EmailSenderServiceImpl.kt [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        return runCatching {
+            javaMailSender.send(generateMailMessage(emailDTO))
+            logger.debug("Email sent successfully")
+            SendEmailResponce(status = 200)
+        }.getOrElse { exception ->
+            logger.warn("An error has occurred sending email", exception)
+            SendEmailResponce(status = 500, cause = exception.message)
+        }
+    }
+    private fun generateMailMessage(emailDTO: EmailDTO): MimeMessage {
```
Comment (ajax-gnatiuk-s) [inline src/main/kotlin/com/example/ajaxproject/service/EmailSenderServiceImpl.kt:27]:
```suggestion
    }
    
    private fun generateMailMessage(emailDTO: EmailDTO): MimeMessage {
```

**Review by ajax-gnatiuk-s — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/example/ajaxproject/pbp/NotificationAnnotationBeanPostProcessor.kt [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+import kotlin.reflect.KClass
+import kotlin.reflect.full.memberFunctions
+
+@Component
+class NotificationAnnotationBeanPostProcessor(
+    @Qualifier("javaMailEmailSenderService") private val emailSenderService: EmailSenderService
+) : BeanPostProcessor {
+
+    @Autowired
+    private lateinit var groupChatRoomRepository: GroupChatRoomRepository
```
Comment (ajax-gnatiuk-s) [inline src/main/kotlin/com/example/ajaxproject/pbp/NotificationAnnotationBeanPostProcessor.kt:28]: why mixed constructor injection and field injection?

Comment (OleksandRomaniuk) [reply src/main/kotlin/com/example/ajaxproject/pbp/NotificationAnnotationBeanPostProcessor.kt]: Really did it really illogically) Rewrote fully in constructor

**Review by ajax-gnatiuk-s — APPROVED** [review]
let's discuss your BPP on next 1-2-1

**Pull request closed without merging**